heat_pump_adoption_modelling/pipeline/supervised_model/data_aggregation.py

```diff
@@ -0,0 +1,429 @@
+    postcode: str
+        Original full postcode.
+
+    Return
+    ---------
+    part_1: str
+        First part of postcode.
+
+    part_2: str
+        Second part of postcode."""
+
+    # Split postcode into two sections
+    try:
+        part_1, part_2 = re.findall(
+            r"([A-Z]{1,2}[0-9]{1,2})\s*([0-9][A-Z]{1,2})", postcode
+        )[0]
+
+    # If not found
+    except IndexError:
+        return None
+
+    return part_1, part_2
+
+
+def get_postcode_area(postcode):
+    """Get postcode area (highest postcode level).
+    For example: YO for YO30 5QW
+
+    Parameters
+    ----------
+    postcode: str
+        Original full postcode.
+
+    Return
+    ---------
+    postcode area : str"""
+
+    # Split postcode into two sections
+    split_pc = split_postcode(postcode)
+
+    if split_pc is None:
+        return np.nan
+
+    # Get first letters
+    part_1, _ = split_pc
+    postcode_area = re.findall(r"([A-Z]+)", part_1)[0]
+
+    return postcode_area
+
+
+def get_postcode_district(postcode):
+    """Get postcode area (2nd highest postcode level).
+    For example: YO30 for YO30 5QW
+
+    Parameters
+    ----------
+    postcode: str
+        Original full postcode.
+
+    Return
+    ---------
+    postcode district : str"""
+
+    # Split postcode into two sections
+    split_pc = split_postcode(postcode)
+
+    if split_postcode is None:
+        return np.nan
+
+    # First section is equal two district
+    part_1, _ = split_pc
+    return part_1
+
+
+def get_postcode_sector(postcode):
+    """Get postcode sector (2nd lowest postcode level).
+    For example: YO30 5 for YO30 5QW
+
+    Parameters
+    ----------
+    postcode: str
+        Original full postcode.
+
+    Return
+    ---------
+    postcode sector : str"""
+
+    # Split postcode into two sections
+    split_pc = split_postcode(postcode)
+
+    if split_postcode is None:
+        return np.nan
+
+    # Get first part + numbers of 2nd part
+    part_1, part_2 = split_pc
+    postcode_sector = part_1 + re.findall(r"[0-9]", part_2)[0]
+    return postcode_sector
+
+
+def get_postcode_levels(df, only_keep=None):
+    """Get 4 different postcode levels: area, district, sector and unit.
+    Optionally, only keep required postcode level.
+
+    Parameters
+    ----------
+    df: pandas.DataFrame
+        Dataframe with column "POSTCODE"
+
+    only_keep: str, efault=None
+        If postcode level is given, the other postcode levels will be removed.
+        For instance, if "POSTCODE_DISTRICT" is given, remove
+        columns "POSTCODE_ADEA/SECTOR/UNIT" from dataframe.
+
+    Return
+    ---------
+    df : pandas.DataFrame with added postcode levels
+    """
+
+    df["POSTCODE_AREA"] = df["POSTCODE"].apply(get_postcode_area)
+    df["POSTCODE_DISTRICT"] = df["POSTCODE"].apply(get_postcode_district)
+    df["POSTCODE_SECTOR"] = df["POSTCODE"].apply(get_postcode_sector)
+    df["POSTCODE_UNIT"] = df["POSTCODE"]
+
+    if only_keep is not None:
```

Comment:
Thanks a lot for all the comments regarding the postcode level splitting. I guess this is what happens when "code evolves naturally" - it gets a bit messy and complicated.
With the help of your suggestions I now implemented a much more readable and slim solution. It even lets you decide whether you want the empty space seperating the two parts, and it can re-insert it if it gets lost. Should come in handy for other projects so I might even split it from the rest of the script.